eth/protocols/eth/dispatcher.go

```diff
@@ -173,10 +196,24 @@ func (p *Peer) dispatchRequests() {
 				pending[req.id] = req
 			}
 
+		case cancelOp := <-p.reqCancel:
+			// Retrieve the pendign request to cancel and short circuit if it
+			// has already been serviced and is not available anymore
+			req := pending[cancelOp.id]
+			if req == nil {
+				cancelOp.fail <- nil
+				continue
+			}
+			// Stop tracking the request
+			delete(pending, cancelOp.id)
+			cancelOp.fail <- nil
+
 		case resOp := <-p.resDispatch:
 			res := resOp.res
 			res.Req = pending[res.id]
-			res.Time = res.recv.Sub(res.Req.sent)
+
```

Comment:
What happens if the response to a `stale` request comes in -- we just drop it, right? I don't quite follow the logic. So we shuffle it over from `pending` to `stale`, but we still don't (immediately) reschedule it, but neither do we handle a late response? 
I must have missed something, but it seems like moving something to `stale` unconditionally means it _will_ time out a bit later, leading to the peer being rejected.

Reply:
Ah, my bad. The `pending` here is dispatcher-local pending, it has no relation to the pending/stales in the downloader.